clients/go/admin/auth_interceptor.go

```diff
@@ -0,0 +1,74 @@
+package admin
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/flyteorg/flyteidl/clients/go/admin/cache"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"google.golang.org/grpc"
+)
+
+// ResetTokenSource will attempt to build a TokenSource given the anonymously available information exposed by the server.
+// Once established, it'll invoke CustomHeaderTokenSource.Reset() on tokenSource to populate it with the appropriate values.
+func ResetTokenSource(ctx context.Context, cfg *Config, tokenCache cache.TokenCache, tokenSource *CustomHeaderTokenSource) error {
+	authMetadataClient, err := InitializeAuthMetadataClient(ctx, cfg)
+	if err != nil {
+		// EngHabu: I don't know why did we use to panic here. Generally speaking, libraries should error on the side of
+		// 	returning errors instead.
+		return fmt.Errorf("failed to initialize Auth Metadata Client. Error: %w", err)
+	}
+
+	tokenSourceProvider, err := NewTokenSourceProvider(ctx, cfg, tokenCache, authMetadataClient)
+	if err != nil {
+		return fmt.Errorf("failed to initialize token source provider. Err: %w", err)
+	}
+
+	clientMetadata, err := authMetadataClient.GetPublicClientConfig(ctx, &service.PublicClientAuthConfigRequest{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch client metadata. Error: %v", err)
+	}
+
+	tSource, err := tokenSourceProvider.GetTokenSource(ctx)
+	if err != nil {
+		return err
+	}
+
+	tokenSource.Reset(tSource, clientMetadata.AuthorizationMetadataKey, cfg.UseInsecureConnection)
+	return nil
+}
+
+// newAuthInterceptor creates a new grpc.UnaryClientInterceptor that forwards the grpc call and inspects the error.
+// If an auth error is detected, it'll panic the process. This is useful for scenarios when auth is enabled after the client
```

Comment:
does this actually panic? do we want it to instead of retrying with a rebuilt client?

Reply:
oh good catch! I guess I was doing some next level DDD (Documentation Driven Development. For a split sec I thought I just coined that but google disappointed me!).

So it does try with the rebuilt client... I initially thought to make it panic to force a restart... but I don't think we need to anymore because it will attempt to rebuild the client on every subsequent call until it manages to get through... so I'll update the docs... plz let me know what you think.